src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,88 @@
+    private final Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        Object clazzImplementationInstance = null;
+        Class<?> clazz = findImplementation(interfaceClazz);
+        validateImplementation(clazz);
+        Field[] declaredFields = clazz.getDeclaredFields();
+        for (Field field: declaredFields) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                Object fieldInstance = getInstance(field.getType());
+                clazzImplementationInstance = createNewInstance(clazz);
+                try {
+                    field.setAccessible(true);
+                    field.set(clazzImplementationInstance, fieldInstance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Can't set field "
+                            + field.getName()
+                            + " for class "
+                            + clazz.getName());
+                }
+            }
+        }
+        if (clazzImplementationInstance == null) {
+            clazzImplementationInstance = createNewInstance(clazz);
+        }
+        return clazzImplementationInstance;
+    }
+
+    private void validateImplementation(Class<?> clazz) {
+        if (clazz == null) {
+            throw new InvalidImplementationException("Implementation can't be null");
+        }
+        if (!clazz.isAnnotationPresent(Component.class)) {
+            throw new InvalidImplementationException(
+                    "Can't create an instance of a non-component class " + clazz.getName());
+        }
+    }
+
+    private Object createNewInstance(Class<?> clazz) {
+        if (instances.containsKey(clazz)) {
+            return instances.get(clazz);
+        }
+        try {
+            Constructor<?> constructor = clazz.getConstructor();
+            Object instance = constructor.newInstance();
+            instances.put(clazz, instance);
+            return instance;
+        } catch (NoSuchMethodException | InstantiationException
+                 | IllegalAccessException | InvocationTargetException e) {
+            throw new RuntimeException(
+                    "Couldn't create an instance of class "
+                    + clazz.getName());
+        }
+    }
+
+    private Class<?> findImplementation(Class<?> interfaceClazz) {
+        if (!interfaceClazz.isInterface()) {
+            return interfaceClazz;
+        }
+        if (interfaceImplementations.size() == 0) {
+            interfaceImplementations.put(FileReaderService.class, FileReaderServiceImpl.class);
+            interfaceImplementations.put(ProductParser.class, ProductParserImpl.class);
+            interfaceImplementations.put(ProductService.class, ProductServiceImpl.class);
+        }
```

Comment:
For the future, it is better to set this data in the "interfaceImplementations" field than to check every time it is empty. you can initialize it with Map.of() method
this method creates an immutable map
``` suggestion
```